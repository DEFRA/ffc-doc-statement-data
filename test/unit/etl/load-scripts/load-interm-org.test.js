const { etlConfig } = require('../../../../app/config')
const db = require('../../../../app/data')
const { loadIntermOrg } = require('../../../../app/etl/load-scripts/load-interm-org')

jest.mock('../../../../app/data', () => ({
  sequelize: {
    query: jest.fn()
  },
  etlStageLog: {
    findAll: jest.fn()
  },
  Sequelize: {
    Op: {
      gt: Symbol('gt')
    }
  }
}))

describe('loadIntermOrg', () => {
  const startDate = '2023-01-01'

  beforeEach(() => {
    db.etlStageLog.findAll.mockClear()
    db.sequelize.query.mockClear()
  })

  test('should throw an error if multiple records are found', async () => {
    db.etlStageLog.findAll.mockResolvedValue([{ file: 'Organization_SFI23/export.csv', idFrom: 1, idTo: 2 }, { file: 'Organization_SFI23/export.csv', idFrom: 3, idTo: 4 }])

    await expect(loadIntermOrg(startDate)).rejects.toThrow(
      `Multiple records found for updates to ${etlConfig.organisation.folder}, expected only one`
    )
  })

  test('should return if no records are found', async () => {
    db.etlStageLog.findAll.mockResolvedValue([])

    await expect(loadIntermOrg(startDate)).resolves.toBeUndefined()
    expect(db.sequelize.query).not.toHaveBeenCalled()
  })

  test('should call sequelize.query with correct SQL and parameters', async () => {
    db.etlStageLog.findAll.mockResolvedValue([{ file: 'Organization_SFI23/export.csv', idFrom: 1, idTo: 2 }])

    await loadIntermOrg(startDate)

    expect(db.sequelize.query).toMatchSnapshot()
  })
})
