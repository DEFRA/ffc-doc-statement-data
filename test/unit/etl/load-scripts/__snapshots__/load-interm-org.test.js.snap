// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadIntermOrg should call sequelize.query with correct SQL and parameters 1`] = `
[MockFunction] {
  "calls": [
    [
      "
 WITH "newData" AS (
    SELECT
      O."sbi",
      A."businessAddress1" AS "addressLine1",
      A."businessAddress2" AS "addressLine2",
      A."businessAddress3" AS "addressLine3",
      A."businessCity" AS "city",
      A."businessCounty" AS "county",
      A."businessPostCode" AS "postcode",
      A."businessEmailAddr" AS "emailAddress",
      A."frn",
      A."businessName" AS "name",
      O."lastUpdatedOn"::date AS "updated",
      O."partyId",
      O."changeType"
    FROM public."etlStageOrganisation" O
    INNER JOIN public."etlStageBusinessAddressContactV" A ON A.sbi = O.sbi
    WHERE O."etlId" BETWEEN 1 AND 2
      
  ),
  "updatedrows" AS (
    UPDATE public."etlIntermOrg" interm
    SET
      "addressLine1" = "newData"."addressLine1",
      "addressLine2" = "newData"."addressLine2",
      "addressLine3" = "newData"."addressLine3",
      city = "newData".city,
      county = "newData".county,
      postcode = "newData".postcode,
      "emailAddress" = "newData"."emailaddress",
      frn = "newData".frn,
      sbi = "newData".sbi,
      "name" = "newData".name,
      updated = "newData".updated,
      "etlInsertedDt" = NOW()
    FROM "newData"
    WHERE "newData"."changeType" = 'UPDATE'
      AND interm."partyId" = "newData"."partyId"
    RETURNING interm."partyId"
  )
  INSERT INTO public."etlIntermOrg" (
    "sbi",
    "addressLine1",
    "addressLine2",
    "addressLine3",
    city,
    county,
    postcode,
    "emailAddress",
    frn,
    "name",
    "partyId",
    updated
  )
  SELECT
    sbi,
    "addressLine1",
    "addressLine2",
    "addressLine3",
    city,
    county,
    postcode,
    "emailAddress",
    frn,
    "name",
    "partyId",
    updated
  FROM "newData"
  WHERE "changeType" = 'INSERT'
    OR ("changeType" = 'UPDATE' AND "partyId" NOT IN (SELECT "partyId" FROM updatedrows));
",
      {
        "raw": true,
        "replacements": {},
        "transaction": undefined,
      },
    ],
    [
      "
 WITH "newData" AS (
    SELECT
      O."sbi",
      A."businessAddress1" AS "addressLine1",
      A."businessAddress2" AS "addressLine2",
      A."businessAddress3" AS "addressLine3",
      A."businessCity" AS "city",
      A."businessCounty" AS "county",
      A."businessPostCode" AS "postcode",
      A."businessEmailAddr" AS "emailAddress",
      A."frn",
      A."businessName" AS "name",
      O."lastUpdatedOn"::date AS "updated",
      O."partyId",
      O."changeType"
    FROM public."etlStageOrganisation" O
    INNER JOIN public."etlStageBusinessAddressContactV" A ON A.sbi = O.sbi
    WHERE O."etlId" BETWEEN 1 AND 2
       AND O."etlId" NOT BETWEEN 1 AND 2
  ),
  "updatedrows" AS (
    UPDATE public."etlIntermOrg" interm
    SET
      "addressLine1" = "newData"."addressLine1",
      "addressLine2" = "newData"."addressLine2",
      "addressLine3" = "newData"."addressLine3",
      city = "newData".city,
      county = "newData".county,
      postcode = "newData".postcode,
      "emailAddress" = "newData".emailaddress,
      frn = "newData".frn,
      sbi = "newData".sbi,
      "name" = "newData".name,
      updated = "newData".updated,
      "etlInsertedDt" = NOW()
    FROM "newData"
    WHERE "newData"."changeType" = 'UPDATE'
      AND interm."partyId" = "newData"."partyId"
    RETURNING interm."partyId"
  )
  INSERT INTO public."etlIntermOrg" (
    "sbi",
    "addressLine1",
    "addressLine2",
    "addressLine3",
    city,
    county,
    postcode,
    "emailAddress",
    frn,
    "name",
    "partyId",
    updated
  )
  SELECT
    sbi,
    "addressLine1",
    "addressLine2",
    "addressLine3",
    city,
    county,
    postcode,
    "emailAddress",
    frn,
    "name",
    "partyId",
    updated
  FROM "newData"
  WHERE "changeType" = 'INSERT'
    OR ("changeType" = 'UPDATE' AND "partyId" NOT IN (SELECT "partyId" FROM updatedrows));
",
      {
        "raw": true,
        "replacements": {},
        "transaction": undefined,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
